pennylane/tape/tape.py

```diff
@@ -930,6 +936,268 @@ def set_parameters(self, params, trainable_only=True):
+        multiple state measurements is not supported.
+        """
+        shape = tuple()
+
+        ret_type = mps[0].return_type
+        shot_vector = device._shot_vector
+
+        # Shot vector was defined
+        if ret_type in (qml.measurements.Expectation, qml.measurements.Variance):
+            num = sum(shottup.copies for shottup in shot_vector)
+            shape = (num, len(mps))
+
+        elif ret_type == qml.measurements.Probability:
+
+            wires_num_set = {len(meas.wires) for meas in mps}
+            same_num_wires = len(wires_num_set) == 1
+            if same_num_wires:
+                # All probability measurements have the same number of
+                # wires, gather the length from the first one
+
+                len_wires = len(mps[0].wires)
+                dim = mps[0]._get_num_basis_states(len_wires, device)
+                shot_copies_sum = sum(s.copies for s in shot_vector)
+                shape = (shot_copies_sum, len(mps), dim)
+
+            else:
+                # There is a varying number of wires that the probability
+                # measurement processes act on
+                # TODO: revisit when issues with this case are resolved
+                raise TapeError(
+                    "Getting the output shape of a tape with multiple probability measurements "
+                    "along with a device that defines a shot vector is not supported."
+                )
+
+        elif ret_type == qml.measurements.Sample:
+            shape = []
+            for shot_val in device.shot_vector:
+                for _ in range(shot_val.copies):
+                    shots = shot_val.shots
+                    if shots != 1:
+                        shape.append(tuple([shots, len(mps)]))
+                    else:
+                        shape.append((len(mps),))
+        return shape
+
+    def shape(self, device):
+        """Produces the output shape of the tape by inspecting its measurements
+        and the device used for execution.
+
+        .. note::
+
+            The computed shape is not stored because the output shape may be
+            dependent on the device used for execution.
+
+        **Example:**
+
+        .. code-block:: python
+
+            import pennylane as qml
+            from pennylane.gradients import param_shift
+            from pennylane import numpy as np
+
+            dev = qml.device("default.qubit", wires=2)
+            a = np.array([0.1, 0.2, 0.3])
+
+            def func(a):
+                qml.RY(a[0], wires=0)
+                qml.RX(a[1], wires=0)
+                qml.RY(a[2], wires=0)
+
+            with qml.tape.QuantumTape() as tape:
+                func(a)
+                qml.state()
+
+            # 1. Execute the tape
+            tape_execute_res = qml.execute(
+                [tape],
+                dev,
+                cache=None,
+                interface="autograd",
+                gradient_fn=param_shift
+            )
```

Comment:
I feel like the minimal example doesn't need to include execution, since we can compute the shape as soon as the tape is constructed? Maybe more complex examples can go further down?

Reply:
Oh, great point! I've removed the execution part. It was meant to serve as a comparison to the QNode output shape.